### PR TITLE
Add `group_cat` functionality

### PR DESCRIPTION
**Pull request by tzuhanchang**
Concatenates the given sequence of `tensors` in the given dimension `dim`. Differ from `torch.cat`, values along the concatenating dimension are grouped according to the indicies defined in the `index` tensors. 

All tensors must either have the same shape (except in the concatenating dimension) or be empty and all index tensors must in the same sequence as in `tensors`.

Examples:
```
>>> x1 = torch.tensor([[0.2716, 0.4233, 0.2658, 0.8284],
...                    [0.3166, 0.0142, 0.1700, 0.2944],
...                    [0.2371, 0.3839, 0.7193, 0.2954],
...                    [0.4100, 0.0012, 0.5114, 0.3353]])
>>> x2 = torch.tensor([[0.3752, 0.5782, 0.7105, 0.4002],
...                    [0.7757, 0.5999, 0.7898, 0.0753]])

>>> x1_index = torch.LongTensor([0,0,1,2])
>>> x2_index = torch.LongTensor([0,2])

>>> scatter_concat([x1,x2], [x1_index,x2_index], 0)
tensor([[0.2716, 0.4233, 0.2658, 0.8284],
              [0.3166, 0.0142, 0.1700, 0.2944],
              [0.3752, 0.5782, 0.7105, 0.4002],
              [0.2371, 0.3839, 0.7193, 0.2954],
              [0.4100, 0.0012, 0.5114, 0.3353],
              [0.7757, 0.5999, 0.7898, 0.0753]])
```

### CODE REVIEW

File: torch_geometric/utils/_scatter.py [hunk 1]
```diff
@@ -283,3 +283,49 @@ def group_argsort(
     ptr = cumsum(count)
 
     return out - ptr[index]
+
+
+def scatter_concat(
```
Comment (rusty1s) [inline torch_geometric/utils/_scatter.py:288]: Can we name this `grouped_cat`?

File: test/utils/test_scatter.py [hunk 2]
```diff
@@ -111,6 +111,28 @@ def test_scatter_argmax(device):
     assert argmax.tolist() == [3, 5, 1, 4, 5, 5]
 
 
+@withCUDA
+def test_scatter_concat(device):
+    x1 = torch.tensor(
+        [[0.2716, 0.4233, 0.2658, 0.8284], [0.3166, 0.0142, 0.1700, 0.2944],
+         [0.2371, 0.3839, 0.7193, 0.2954], [0.4100, 0.0012, 0.5114, 0.3353]],
+        device=device)
```
Comment (rusty1s) [inline test/utils/test_scatter.py:119]: test with `torch.randn` should be better IMO.

**Review by rusty1s — APPROVED** [review]
